pkg/controller/endpoint/endpoints_controller_cache.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoint
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+// EndpointsControllerCache is a util that manages a local copy of the state of pods and services
+// for purpose of endpoints object syncing.
+//
+// The purpose of this util is to create a single source of truth that is thread-safe and avoids
+// potential race-conditions in the endpoints-controller. In addition it allows to easily determine
+// which object (pod/service) changes resulted in the endpoints object change, which is crucial in
+// implementing the network programming latency metric.
+//
+// The cache should be initialized via pod and service listers and kept up-to-date via the
+// (Add|Update|Delete)(Pod|Service) methods called in the EndpointsController's event handlers.
+// The Initialize method together with the event handlers method are guarded by the same mutex which
+// guarantees that no information is ever missed and all events are processed in the serial order.
+//
+// The cache doesn't store full objects, it stores only references to minimize the memory
+// consumption.
+type endpointsControllerCache struct {
+	pods     map[ /* namespace */ string]map[ /* name */ string]*v1.Pod
```

Comment:
nit: I would prefer to have the namespace -> name -> Pod mapping be written in a commen trather than be inlined.